coordinator/src/trade/mod.rs

```diff
@@ -740,18 +959,50 @@ impl TradeExecutor {
                     compute_relative_contracts(contracts, &trade_params.direction)
                 };
 
-                if position_contracts + trade_contracts == Decimal::ZERO {
-                    TradeAction::ClosePosition {
+                let average_execution_price = trade_params.filled_with.average_execution_price();
+
+                match (position_contracts, trade_contracts) {
+                    // If they cancel out, we are closing the position.
+                    (p, t) if (p + t).is_zero() => TradeAction::ClosePosition {
+                        channel_id,
+                        position: Box::new(position),
+                    },
+                    // If they signs are the same, we are increasing the position.
+                    (p, t) if (p * t).is_sign_positive() => TradeAction::ResizePosition {
                         channel_id,
                         position: Box::new(position),
+                        resize_action: ResizeAction::Increase {
+                            contracts: t.abs(),
+                            average_execution_price,
+                        },
+                    },
+                    // If they signs are the different and the trade contracts are lower than the
+                    // position contracts, we are decreasing the position.
+                    (p, t) if (p * t).is_sign_negative() && t.abs() < p.abs() => {
+                        TradeAction::ResizePosition {
+                            channel_id,
+                            position: Box::new(position),
+                            resize_action: ResizeAction::Decrease {
+                                contracts: t.abs(),
+                                average_execution_price,
+                            },
+                        }
+                    }
+                    // If they signs are the different and the trade contracts are greater than the
+                    // position contracts, we are changing position direction.
+                    (p, t) => {
+                        let contracts_new_direction = t.abs() - p.abs();
+                        let contracts_new_direction = contracts_new_direction.abs();
+
+                        TradeAction::ResizePosition {
+                            channel_id,
+                            position: Box::new(position),
+                            resize_action: ResizeAction::ChangeDirection {
+                                contracts_new_direction,
+                                average_execution_price,
+                            },
+                        }
                     }
-                } else {
-                    ensure!(
```

Comment:
🔧 this got lost and should be checked for every resize action.

Reply:
Good catch, thanks!